Add Handling Req Body Guide (#808)
zeit/docs: pages/guides/handling-node-request-body-with-micro.mdx

import Guide from '~/components/layout/guide'
import { TerminalInput } from '~/components/text/terminal'
import { Image } from '~/components/media'
import Caption from '~/components/text/caption'
import { InlineCode } from '~/components/text/code'
import Card from '~/components/card'
import { Code } from '~/components/text/code'
import Note from '~/components/text/note'
import Request from '~/components/api/request'
import { GenericLink } from '~/components/text/link'

export const meta = {
  title: 'Handling Node.js Request Bodies With Micro',
  description:
    'Learn how to parse Node.js request bodies with Micro for use inside serverless functions.',
  published: '2019-05-01T09:03:14.819Z',
  authors: ['msweeneydev'],
  url: '/guides/handling-node-request-body-with-micro',
  editUrl: 'pages/guides/handling-node-request-body-with-micro.mdx',
  lastEdited: '2019-05-07T00:16:43.000Z'
}

> In this guide, we will show you how to parse a [Node.js request body](https://nodejs.org/en/docs/guides/anatomy-of-an-http-transaction/), for use inside a [serverless function](/docs/v2/deployments/concepts/lambdas/), without requirinq a framework such as Express.

Reviewer: ```suggestion
In this guide, we will show you how to parse a [Node.js request body](https://nodejs.org/en/docs/guides/anatomy-of-an-http-transaction/), for use inside a [serverless function](/docs/v2/deployments/concepts/lambdas/), without requiring a framework such as Express.
```

Reviewer: Is micro not a framework? I'm not sure about the terminology there.

> This guide assumes the request is sent with a [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) of [application/json](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/JSON). However, many more content types can be parsed using the [`micro` module](https://github.com/zeit/micro).

Reviewer: ```suggestion
This guide assumes the request is sent with a [`Content-Type`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) of [`application/json`](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/JSON). However, many more content types can be parsed using the [`micro` module](https://github.com/zeit/micro).
```

## Step 1: Creating the Function

> To illustrate the parsing of the request body, let's take the simple function below as an example:

Reviewer: ```suggestion
To illustrate the parsing of a request body, let's take the simple function below as an example:
```

```js
const { json } = require('micro')

module.exports = async (req, res) => {
  const body = await json(req)
  res.end(`Hello ${body.name}, you just parsed the request body!`)
}
```

<Caption>
  An example of how to parse a request body using{' '}

>   <GenericLink href="https://nodejs.org/dist/latest-v12.x/docs/api/">

Reviewer: Linking to the v12 docs might be a bad idea since Now uses v8.x

Author: Changed to v8 docs.

    Node.js
  </GenericLink>{' '}
  and <GenericLink href="https://github.com/zeit/micro">Micro</GenericLink>.
</Caption>

> This function takes a POST request, parses the body and uses it in the response.

Reviewer: ```suggestion
This function takes a POST request, parses the body, and uses data from the body in the response.
```

Reviewer: Perhaps we can link the "response" to what requests and responses are in terms of @now/node in the docs.

> The key part here is line 3, for those used to working with Express, it should be similar to accessing `req.body`.

Reviewer: This makes me think we should add line numbers to code blocks. cc @evilrabbit 

```suggestion
The key part here is line 4. For those used to working with Express, it should be similar to accessing `req.body`.
```

Including the newline, it's line 4, no?

> The `json(req)` function returns the request body in the JSON format, fully parsed and ready to work with.

Reviewer: ```suggestion
`json`, imported from micro on line 1, is a function that returns the request body in the JSON format, fully parsed and ready to access through your function as you would with any object.
```
Link wiki/mdn for JSON? All about references.

This is illustrated in the response, as we access the JSON object to use the `name` property that was sent in the initial request.

## Step 2: Sending the Request

To verify that the JSON is being parsed correctly, make a POST request using [curl](https://curl.haxx.se/) by executing the below code inside your terminal:

> <Request

Reviewer: Nice use of Request outside of API!

  method="POST"
  url="https://req-body-micro.now.sh/"
  headers={{
    'Content-Type': 'application/json'
  }}
  body={{
    name: 'Reader'
  }}
/>
<Caption>
    Making a POST request using <GenericLink href="https://curl.haxx.se/">curl</GenericLink>.
</Caption>

> The endpoint contains only the code shown in [step 1](#step-1:-setting-up-the-function) with nothing else added to it. You will receive a response like the following:

Reviewer: This isn't true, right? It includes config with now.json. I'd think that I could deploy without config if I read this for the first time.

Author: Added link to the the configuration section showing the use of a `now.json` file.

<Code>Hello Reader, you just parsed the request body!</Code>
<Caption>
    An example response from making a POST request.
</Caption>

## Step 3: Understanding Why this Works

When [Node.js](https://nodejs.org/dist/latest-v12.x/docs/api/) receives a request, the [body](https://nodejs.org/en/docs/guides/anatomy-of-an-http-transaction/#request-body) is in the format of a [ReadableStream](https://nodejs.org/api/stream.html#stream_class_stream_readable).

To get the data from the stream, you need to listen to its `data` and `end` events. To do this requires a few lines of code which you would much rather not have to repeat.

By using [`micro`](https://github.com/zeit/micro), parsing [request bodies](https://nodejs.org/en/docs/guides/anatomy-of-an-http-transaction/#request-body) is concise and straightforward as it handles everything for you, all you need to do is specify the format you want to receive the data in.

## More Resources

> For more information on working with [Now for GitLab](/gitlab), please refer to [the Now for GitLab documentation](/docs/v2/integrations/now-for-gitlab).

Reviewer: Let's change this to "more information on deploying a Node.js API with Now" or something?

Author: Added link to @now/node builder.
